chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -932,36 +998,41 @@ impl<'a, S: BlockchainStorageWrite, V: TransactionVerificationStrategy> Chainsta
 
         // Disconnect the current chain if it is not a genesis
         if let GenBlockId::Block(best_block_id) = best_block_id.classify(self.chain_config) {
-            let mainchain_tip = self
-                .get_block_index(&best_block_id)
-                .map_err(|err| BlockError::BlockIndexQueryError(err, best_block_id.into()))
-                .log_err()?
-                .ok_or(BlockError::InvariantErrorBestBlockIndexNotFound(
-                    best_block_id.into(),
-                ))
-                .log_err()?;
-
             // Disconnect blocks
-            self.disconnect_until(&mainchain_tip, common_ancestor_id).log_err()?;
+            self.disconnect_until(&best_block_id, common_ancestor_id).log_err()?;
         }
 
         // Connect the new chain
         for block_index in new_chain {
             let block: WithId<Block> = self
                 .get_block_from_index(&block_index)
                 .map_err(BlockError::StorageError)
-                .and_then(|opt| {
-                    opt.ok_or(BlockError::InvariantErrorBlockNotFoundAfterConnect(
-                        (*block_index.block_id()).into(),
-                    ))
+                .log_err()?
+                .ok_or_else(|| {
+                    // Note: missing block data is an expected situation; it can happen if
```

Comment:
One solution that occurred to me is to just delete the block index altogether in that case. But we don't have to do that right now.

Reply:
I've mentioned it in the comment for now.